### PR TITLE
Add Quark.run and Session support

### PR DESCRIPTION
**Pull request by paulofaria**
# Summary

This PR addresses #19 adding native session support to `Quark`. And most importantly it adds `Quark.run` as the entry point of the application.

- [x] Adds `Quark.run` for configuration
- [x] Makes `FileResponder` be the default fallback for `Router` and `Resource` for static file serving
- [x] Embeds `SessionMiddleware` into `Server` through `Quark.run`
- [x] Moves `ContentNegotiationMiddleware` to `Server` through `Quark.run`

# Design decisions

Moving middleware out of the `Router` and `Resource` into an appropriate place for configuration (`Quark.run`) improves modularity.  `Quark.run` will be the entry point of the application. Currently it is responsible to parse the executable arguments and expose them for user configuration and also to configure built-in middleware which are applied to the `Server`.

```swift
Quark.run { parameters in
    let store = InMemoryStore()
    let app = Application(store: store)
    let router = Router(app: app)
    return router
}
```

In the future `parameters` will also contain information parsed from a configuration file and from environment variables. If you define a parameter in more than one medium the priority will be:

1 - Environment variables
2 - Executable arguments
3 - Configuration file

Simple example. Imagine you have a `.quark.yml` file like this:

```yaml
database-uri: postgres://user:password@123.123.123.123:5432/todos
```

And you run the app like this:

```sh
app --database-uri postgres://user:password@localhost:5432/todos
```

The executable argument parameter will override the configuration file parameter. And if you have an environment variable with the format `DATABASE_URI` with some other value.. it will override everything else.

An advantage of `Quark.run` is that it has a nice parity with `quark run`, the command from Quark's command line tool which runs the application.

### CODE REVIEW

File: Sources/Quark/Quark+Venice/Quark+Venice.swift [hunk 1]
```diff
@@ -57,3 +51,9 @@ extension FileResponder {
 //         return File.self
 //     }
 // }
+//
+// extension Router {
+//     public var file: FileProtocol.Type {
```
Comment (Danappelxx) [inline Sources/Quark/Quark+Venice/Quark+Venice.swift]: Is this supposed to be here?

**Comment by dreymonde** [comment]
First, I support the concept of `Quark.run`. It’s nice to have server setup logic contained in one place, even in this small format.
Second, I don’t like the problem of duplicating parameters. Maybe we can make `parameters` an entry point to more specific parameters domains, like these:
```swift
let dbParams = parameters[.database] as? DatabaseParameters
let serverParams = parameters[.server] as? ServerParameters
let envs = parameters[.environment] as? EnvironmentVariables
```

Or with extensions:

```swift
let dbParams = parameters.database
let serverParams = paramaters.server
let ends = parameters.environment
```

And so on.

**Comment by paulofaria** [comment]
> Second, I don’t like the problem of duplicating parameters.

What do you mean by this? About the suggestions yeah. I already have basic support for something like this:

```swift
let dbParams = parameters.database
let serverParams = paramaters.server
let ends = parameters.environment
```

**Comment by dreymonde** [comment]
@paulofaria maybe I didn’t fully understand this paragraph:
> In the future parameters will also contain information parsed from a configuration file and from environment variables. If a parameter is duplicated the priority will be:

> 1 - Environment variables
2 - Executable arguments
3 - Configuration file

**Comment by paulofaria** [comment]
Simple example. Imagine you have a `.quark.yml` file like this:

```yaml
database-uri: postgres://user:password@123.123.123.123:5432/todos
```

And you run the app like this:

```
app --database-uri postgres://user:password@localhost:5432/todos
```

The executable argument will override the configuration file. And if you have an environment variable with the format `DATABASE_URI` with some value.. it will override everything.

**Comment by dreymonde** [comment]
Yes, I didn’t understand that correctly 😄 
That totally makes then.

**Comment by Danappelxx** [comment]
ahh you skipped code review. I'll leave my comments anyway

File: Sources/Quark/Quark+Venice/Quark+Venice.swift [hunk 2]
```diff
@@ -57,3 +55,9 @@ extension FileResponder {
...
+//         return File.self
+//     }
+// }
```
Comment (Danappelxx) [inline Sources/Quark/Quark+Venice/Quark+Venice.swift]: line break at EOF

**Comment by paulofaria** [comment]
you said you would review today I thought your review was no comments haha. yeah don't worry. nothing's final.

File: Sources/Quark/Quark/HTTP/Router/Routes.swift [hunk 3]
```diff
@@ -0,0 +1,1199 @@
...
+            }
+
+            return try respond(request, a, b, c, d, content)
+        }
+
+        addRoute(method: method, path: path, middleware: [contentMapper] + middleware, responder: responder)
+    }
+
+    private func parseParameters(keys: [String], pathParameters: [String: String], count: Int) throws -> [String] {
+        //        Todo: Fix bug in TrieRouteMatcher https://github.com/Zewo/Zewo/issues/122
```
Comment (Danappelxx) [inline Sources/Quark/Quark/HTTP/Router/Routes.swift]: This has already been fixed a while ago

File: Sources/Quark/Quark/Middleware/LogMiddleware/LogMiddleware.swift [hunk 4]
```diff
@@ -1,24 +1,20 @@
...
+        message += "Request:\n\n"
         message += debug ? "\(request.debugDescription)\n" : "\(request)\n"
         message += "--------------------------------------------------------------------------------\n"
-        message += "Response:\n"
+        message += "Response:\n\n"
         message += debug ? "\(response.debugDescription)\n" : "\(response)\n"
-        message += "================================================================================\n\n"
-        logger.log(message)
+        message += "================================================================================\n"
+        print(message)
```
Comment (Danappelxx) [inline Sources/Quark/Quark/Middleware/LogMiddleware/LogMiddleware.swift]: Why did you remove logger injection? This makes this middleware pretty useless. You could have just made a custom `print` logger and defaulted to it...

File: Sources/Quark/Quark/Middleware/SessionMiddleware/SessionMiddleware.swift [hunk 5]
```diff
@@ -0,0 +1,75 @@
+var count = 0
+
+private func uuid() -> String {
+    count += 1
+    return String(count)
```
Comment (Danappelxx) [inline Sources/Quark/Quark/Middleware/SessionMiddleware/SessionMiddleware.swift]: Why not use a real uuid? This is not safe.

File: Sources/Quark/Quark/Middleware/SessionMiddleware/SessionMiddleware.swift [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+public final class SessionMiddleware: Middleware {
+    public static let cookieName = "zewo-session"
```
Comment (Danappelxx) [inline Sources/Quark/Quark/Middleware/SessionMiddleware/SessionMiddleware.swift]: Maybe call it `quark-session` now?

File: Sources/Quark/Venice/File/File.swift [hunk 7]
```diff
@@ -201,14 +201,25 @@ extension File {
...
+        do {
+            try ensureLastOperationSucceeded()
+        } catch {
+            fatalError("Error: \(error)")
+        }
+
         return String(cString: workingDirectory!)
     }
 
+    public static func changeWorkingDirectory(path: String) throws {
```
Comment (Danappelxx) [inline Sources/Quark/Venice/File/File.swift]: maybe `changeWorkingDirectory(to:)` instead?

**Comment by Danappelxx** [comment]
Gotcha. Either way, our guidelines say someone else should merge 😛 .

If you could address the comments on the same branch and submit another pr, that would be fantastic 👍
